### PR TITLE
fix(testing): reduce flakiness in extension e2e tests

### LINKED ISSUE #5472: [Testing]: fix VSCODE_PROXY_URI flakiness
## What is your suggestion?

To improve the reliability of the `extensions.test.ts` test should have access to VSCODE_PROXY_URI which has been failing more than usual in CI lately.

## Why do you want this feature?

More confidence in e2e tests and faster working cycle.

## Are there any workarounds to get this functionality today?

Re-run the tests manually when they fail.

## Are you interested in submitting a PR for this?

Yes.

### PR DESCRIPTION
**Pull request by jsjoeio**
This PR modifies our e2e test to reduce flakiness in the extension tests.

Fixes #5472

### CODE REVIEW

File: test/e2e/models/CodeServer.ts [hunk 1]
```diff
@@ -282,6 +282,32 @@ export class CodeServerPage {
     this.codeServer.logger.debug("Editor is ready!")
   }
 
+  /**
+   * Checks if the test extension has loaded
+   *
+   * Reload until both checks pass
+   */
+  async reloadUntilTestExtensionIsLoaded() {
+    this.codeServer.logger.debug("Waiting for test extenion to load...")
```
Comment (code-asher) [inline test/e2e/models/CodeServer.ts:291]:
```suggestion
    this.codeServer.logger.debug("Waiting for test extension to load...")
```

File: test/e2e/models/CodeServer.ts [hunk 2]
```diff
@@ -282,6 +282,32 @@ export class CodeServerPage {
...
+    while (!extensionIsLoaded) {
+      await this.page.waitForLoadState("load")
+      // Give it an extra second just in case it's feeling extra slow
+      await this.page.waitForTimeout(1000)
+      reloadCount += 1
+      if (await this.isTestExtensionLoaded()) {
+        this.codeServer.logger.debug(`test extension loaded after ${reloadCount} reloads`)
+        break
+      }
+      await this.reloadUntilEditorIsReady()
```
Comment (code-asher) [inline test/e2e/models/CodeServer.ts:305]: We already `await` this on every page load so we can probably omit it.

File: test/e2e/models/CodeServer.ts [hunk 3]
```diff
@@ -296,6 +322,21 @@ export class CodeServerPage {
...
   }
 
+  /**
+   * Checks if the test extension loaded
+   */
+  async isTestExtensionLoaded() {
+    const selector = "text=test extension loaded"
+    this.codeServer.logger.debug("Waiting for test extension to load...")
+
+    await this.page.waitForLoadState("load")
```
Comment (code-asher) [inline test/e2e/models/CodeServer.ts:332]: We already `await` the load event on every page load (via `reloadUntilEditorIsReady`) so we can probably omit.

File: test/e2e/extensions.test.ts [hunk 4]
```diff
@@ -8,12 +8,14 @@ function runTestExtensionTests() {
   test("should have access to VSCODE_PROXY_URI", async ({ codeServerPage }) => {
     const address = await getMaybeProxiedCodeServer(codeServerPage)
 
+    await codeServerPage.reloadUntilTestExtensionIsLoaded()
```
Comment (code-asher) [inline test/e2e/extensions.test.ts:11]: Were you seeing the extension not get loaded at all on the first attempt or was it just taking a bit of time to load?

If the latter, my thinking is that we risk reloading the page multiple times if the extension host is a bit slow when really we just need to wait for the extension to get activated.  Something like:

```
this.codeServer.logger.debug("Waiting for test extension to load...")
await this.page.waitForSelector("text=test extension loaded")
```

If we do this you can ignore all my other comments as they will be obsolete.  :stuck_out_tongue:

File: test/e2e/models/CodeServer.ts [hunk 5]
```diff
@@ -282,6 +282,32 @@ export class CodeServerPage {
...
+  async reloadUntilTestExtensionIsLoaded() {
+    this.codeServer.logger.debug("Waiting for test extenion to load...")
+
+    const extensionIsLoaded = await this.isTestExtensionLoaded()
+    let reloadCount = 0
+
+    while (!extensionIsLoaded) {
+      await this.page.waitForLoadState("load")
+      // Give it an extra second just in case it's feeling extra slow
+      await this.page.waitForTimeout(1000)
```
Comment (code-asher) [inline test/e2e/models/CodeServer.ts:299]: We do the same in `reloadUntilEditorIsReady` so I wonder if we can omit?

File: test/e2e/models/CodeServer.ts [hunk 6]
```diff
@@ -282,6 +282,32 @@ export class CodeServerPage {
...
+      reloadCount += 1
+      if (await this.isTestExtensionLoaded()) {
+        this.codeServer.logger.debug(`test extension loaded after ${reloadCount} reloads`)
+        break
+      }
+      await this.reloadUntilEditorIsReady()
+    }
+
+    this.codeServer.logger.debug("Test extenion has loaded!")
```
Comment (code-asher) [inline test/e2e/models/CodeServer.ts:308]:
```suggestion
    this.codeServer.logger.debug("Test extension has loaded!")
```

File: test/e2e/models/CodeServer.ts [hunk 7]
```diff
@@ -282,6 +282,32 @@ export class CodeServerPage {
...
+    const extensionIsLoaded = await this.isTestExtensionLoaded()
+    let reloadCount = 0
+
+    while (!extensionIsLoaded) {
+      await this.page.waitForLoadState("load")
+      // Give it an extra second just in case it's feeling extra slow
+      await this.page.waitForTimeout(1000)
+      reloadCount += 1
+      if (await this.isTestExtensionLoaded()) {
+        this.codeServer.logger.debug(`test extension loaded after ${reloadCount} reloads`)
```
Comment (code-asher) [inline test/e2e/models/CodeServer.ts:302]: Tiny nit but I recommend using `plural` so we can get `1 reload` vs `2 reloads` etc.

**Review by code-asher — APPROVED** [review]

**Comment by code-asher** [comment]
Great sleuthing on the flakiness!!

File: test/e2e/models/CodeServer.ts (hunk 3 above)
Comment (jsjoeio) [reply test/e2e/models/CodeServer.ts:332]: Good call! I'll remove this.

File: test/e2e/extensions.test.ts [hunk 8]
```diff
@@ -5,15 +5,18 @@ import { describe, test, expect } from "./baseFixture"
 
 function runTestExtensionTests() {
   // This will only work if the test extension is loaded into code-server.
+  // fake commit
```
Comment (jsjoeio) [inline test/e2e/extensions.test.ts:8]:
```suggestion
```

**Review by code-asher — APPROVED** [review]

**Pull request merged**